trustmanager/grpckeystore/grpckeystore.proto

```diff
@@ -0,0 +1,123 @@
+syntax = "proto3";
+
+package grpckeystore;
+
+import "google/protobuf/Empty.proto";
+
+service GRPCKeyStore {
+
+    // NewKey allows they GRPC keystore to provide keypair
+    // used for keystores that generate and never divulge private key
+    rpc NewKey (NewKeyReq) returns (NewKeyRsp) {
+    }
+
+    // Add a key to the GRPC keystore
+    // AddKey maps to KeyStore.AddKey()
+    rpc AddKey (AddKeyReq) returns (AddKeyRsp) {
+    }
+
+    // Get a previously stored key from the GRPC keystore
+    // GetKey maps to KeyStore.GetKey()
+    rpc GetKey (GetKeyReq) returns (GetKeyRsp) {
+    }
+
+    // Get specific key information from the GRPC keystore
+    // GetKeyInfo maps to KeyStore.GetKeyInfo()
+    rpc GetKeyInfo (GetKeyInfoReq) returns (GetKeyInfoRsp) {
+    }
+
+    // List keys available in the GRPC keystore
+    // ListKeys maps to KeyStore.ListKeys()
+    rpc ListKeys (google.protobuf.Empty) returns (ListKeysRsp) {
+    }
+
+    // Remove a key from the keystore
+    // RemoveKey maps to KeyStore.RemoveKey()
+    rpc RemoveKey (RemoveKeyReq) returns (RemoveKeyRsp) {
+    }
+
+    // Sign a message with the specified key
+    // Sign maps to PrivateKey.Sign()
+    rpc Sign (SignReq) returns (SignRsp) {
+    }
+}
+
+
+message NewKeyReq {
+    string Gun = 1;          // see notary/tuf/data/types.go -> GUN type
+    string Role = 2;         // see notary/tuf/data/types.go -> TUFTypes
+    string Algorithm = 3;    // see notary/tuf/data/types.go -> Key types
+    string SignatureAlgorithm = 4; // see notary/tuf/data/types.go -> Signature types
+}
+
+message NewKeyRsp {
+    string KeyId = 1;       // for NewKey, KeyId is generated by remote store
+    string PublicKey = 2;   // Public Key in PEM format, Private Key is never shared
+    string DebugMsg = 3;    // Optional informational string returned from remote
+}
+
+message AddKeyReq {
+    string KeyId = 1;
```

Comment:
With AddKey, I think there may be interfaces where KeyId cannot be specified?

Reply:
I might need some help understanding the case you are concerned about.  From what I can gather from the current yubi and generic keystore implementations, they seem to take for granted that AddKey() provides the keyID via privKey.ID().

Reply:
As the key ID is the TUF key ID, we'd probably always want the client to specify it, otherwise the server would have to understand how to generate a TUF key ID. 

That being said, the ID would need to be generated by the GRPC client wrapper as the `KeyStore.AddKey` method doesn't take an ID. That's how all the other concrete `KeyStore` implementers work though so nothing strange there (https://github.com/theupdateframework/notary/blob/master/trustmanager/keystore.go#L105).

Reply:
OK sounds like we want to keep KeyID in the AddKeyReq message, which will be set by the client wrapper e.g. keyID := privKey.ID() .  Sounds like we should remove keyID from NewKeyResp, since that would have required that the HSM side to generate a keyID, which we want to avoid.

Reply:
this will mean AddKey will be required even when NewKey was used (so that the KeyID can be communicated).   Also means that when NewKey was used,  AddKey will not supply a PrivateKey.

Reply:
The key ID is generated from the public key so we don't need to use AddKey with NewKey. The fact NewKey returns the public key is sufficient.

PrivateKey.ID() is actually inherited from the PublicKey interface being embedded in PrivateKey (https://github.com/theupdateframework/notary/blob/master/tuf/data/keys.go#L29)

Reply:
If we go with notary (grpc client) always generating the KeyID (even in the NewKey case), the KeyID still needs to get over to the HSM (grpc server) somehow.  Was thinking AddKey would be they way to send it over.  Otherwise how would it get there?

Reply:
Ahhh, I see what you're saying. It's an interesting problem. What would be included in the AddKey message to enable the HSM to associate the KeyID with the correct key? We wouldn't have the private key if it had been generated on the HSM via NewKey.

Reply:
Looking at the YubiHSM API for example in https://developers.yubico.com/YubiHSM2/Commands/Generate_Asymmetric_Key.html it needs a key ID thats specific to the HSM (ie 2 bytes), which is a spare ID, so you would not want to specify it (you want the GRPC endpoint to pick a spare one), but it does need to be returned so you can pass it back to use the key. Unless the API endpoint keeps an association between Notary IDs and HSM key ids.

Reply:
I think we have several choices here... I can see any of the following ideas working, but they all have pluses and minuses:
1) idea 1: have the GRPC server generate the KeyID and return it in the NewKeyRsp.  That was my original thought with this .proto.  However, that comes with the downside of having to tach the GRPC server how to generate KeyIDs, as you noted.  In this model, an AddKey would not follow a NewKey, since only one is needed.
2) idea 2: If we want to make sure the notary client generated the KeyID, we could require an AddKey sequence even when NewKey was used, so that the KeyID can be sent over.  AddKey would contain the public key that was returned in NewKey so the server side would know what key we are talking about, but the private key would be empty in this case.
3) idea 3: To avoid overloading AddKey as shown idea 2, but still want have the notary client generate the KeyID, we could add an additional exchange -- something like like UseKey.  UseKey would always follow NewKey, and would basically indicate that the notary client accepted the NewKey and would provide the KeyID in UseKey.

Reply:
Given we're implementing a GRPC service, I think it's reasonable to expect there is some type of server wrapping the HSM (let me know if you disagree).

In that case, we don't need to be bound to the ID scheme of the HSM, the wrapping server can implement a mapping from TUF IDs to the HSM's ID scheme.

With those assumptions, I'd prefer option 3 rather than overloading `AddKey`. Something like `Associate(id, publicKey)`

Reply:
@justincormack You bring up a really good point.  I like the idea of adding a server side correlator such as  "RemoteID"  that the server provides on NewKeyRsp, AddKeyRsp, and ListKeysRsp, and that the the client will echo back on GetKeyReq, GetKeyInfoReq, RemoteKeyReq, and SignReq.  That way, the remote store can correlate to keys using whatever it's preferred method happens to be.  However, that's really solves a different issue than the one were talking about here.  To make ListKeys work across a GRPC interface it's looking like the remote server still must store the KeyID, GUN and role alongside every key pair it holds.